### PR TITLE
Switch to pdoc

### PR DESCRIPTION
**Pull request by davfsa**
### Summary
Switch to pdoc with new theme style and better mobile compatibility

This branch is not 100% functional to be merged into master, as it relies a bit on having multiple documentation versions. The idea is to have this merged into a separate branch and then it can be worked on from there, as this is already really big

### Related issues
Solves all points in #808

<details>
<summary>Screenshots</summary>

#### Desktop
image
image

#### Mobile
image
Screen Shot 2021-10-19 at 22 07 28
Screen Shot 2021-10-19 at 22 07 51
</details>

### CODE REVIEW

**Comment by davfsa** [comment]
Fixed failing pipelines and added some screenshots to the initial comment, in case anybody is interested :)

File: hikari/impl/bot.py [hunk 1]
```diff
@@ -213,6 +212,15 @@ class GatewayBot(traits.GatewayBotAware):
     ```
     """
 
+    shards: typing.Mapping[int, gateway_shard.GatewayShard]
+    """Mapping of shards in this application instance.
+
+    Each shard ID is mapped to the corresponding shard instance.
+
+    If the application has not started, it is acceptable to assume the
```
Comment (FasterSpeeding) [inline hikari/impl/bot.py:220]: Should this doc string not be being inherited from traits.ShardAware

File: hikari/impl/event_manager.py [hunk 2]
```diff
@@ -88,7 +88,7 @@ def __init__(
         super().__init__(event_factory=event_factory, intents=intents)
 
     async def on_ready(self, shard: gateway_shard.GatewayShard, payload: data_binding.JSONObject) -> None:
-        """See https://discord.com/developers/docs/topics/gateway#ready for more info."""
+        """See <<https://discord.com/developers/docs/topics/gateway#ready>> for more info."""
```
Comment (FasterSpeeding) [inline hikari/impl/event_manager.py:91]: accidental double brackets

File: pipelines/utils.nox.py [hunk 3]
```diff
@@ -25,29 +25,38 @@
...
+
+FILES_TO_DELETE = [
+    ".coverage",
     "package-lock.json",
 ]
 
 
 @nox.session(reuse_venv=False)
-def purge(_: nox.Session) -> None:
+def purge(session: nox.Session) -> None:
```
Comment (FasterSpeeding) [inline pipelines/utils.nox.py:46]: this seems a bit out of scope

File: hikari/internal/deprecation.py [hunk 4]
```diff
@@ -49,60 +47,25 @@ def warn_deprecated(
...
         The object that is deprecated.
+    reason: str
+        The reason for the deprecation.
 
     Other Parameters
     ----------------
-    version: typing.Optional[str]
-        If specified, the version it will be removed in.
-    alternative: typing.Optional[str]
-        If specified, the alternative to use.
```
Comment (FasterSpeeding) [inline hikari/internal/deprecation.py:58]: why were these args removed and what do these changes have to do with pdoc

File: hikari/internal/deprecation.py [hunk 5]
```diff
@@ -49,60 +47,25 @@ def warn_deprecated(
...
+        action = ("Call to", "function/method")
 
-    if alternative is not None:
-        message += f" You can use '{alternative}' instead."
-
-    warnings.warn(message, category=DeprecationWarning, stacklevel=stack_level)
-
-
-def deprecated(
-    version: typing.Optional[str] = None, alternative: typing.Optional[str] = None
```
Comment (FasterSpeeding) [inline hikari/internal/deprecation.py:75]: why was this removed

File: hikari/impl/bot.py (hunk 1 above)
Comment (davfsa) [reply hikari/impl/bot.py:220]: It doesnt seem to be, most probably since it becomes an attribute instead of a property

File: pipelines/utils.nox.py (hunk 3 above)
Comment (davfsa) [reply pipelines/utils.nox.py:46]: I know, but it was annoying me quite a bit when it didnt work correctly

File: hikari/internal/deprecation.py (hunk 4 above)
Comment (davfsa) [reply hikari/internal/deprecation.py:58]: _This was explained on Discord, but I will put it here for context_

The decorator seemed to completely discard the docstring, and I couldn't get it to work, which is why I make it a function only instead.

Will try to get this to work when I have time

Comment (FasterSpeeding) [reply hikari/internal/deprecation.py:58]: What about the args though

File: pipelines/utils.nox.py (hunk 3 above)
Comment (FasterSpeeding) [reply pipelines/utils.nox.py:46]: wouldn't it be less annoying if you fixed it sooner by putting it in a pr which can be immediately pred to master instead of tagging it along with an unrelated massive change

File: hikari/impl/bot.py (hunk 1 above)
Comment (FasterSpeeding) [reply hikari/impl/bot.py:220]: Well that shouldn't be an issue if you don't move where its being type hinting it for no reason but also cannot reproduce that behaviour

Comment (davfsa) [reply hikari/impl/bot.py:220]: I am still able to reproduce it unfortunately: 
image

I got absolutely no idea why this happens

Comment (FasterSpeeding) [reply hikari/impl/bot.py:220]: It shouldn't be an issue if its being type hinted in the `__init__` though since pdoc just doesn't introspect those

Comment (davfsa) [reply hikari/impl/bot.py:220]: Nope, even if I take the typehint out it still loses the docstring for some reason. I really havent been able to figure out

**Review by FasterSpeeding — APPROVED** [review]

File: pipelines/config.py [hunk 6]
```diff
@@ -95,6 +95,5 @@
     *(f for f in _os.listdir(".") if _os.path.isfile(f) and f.endswith(REFORMATTING_FILE_EXTS)),
     ".github",
     "pages",
-    "docs",
```
Comment (davfsa) [inline pipelines/config.py:98]: Re-add this and use `--ignore-words-list="nd"` when running codespell

Comment (FasterSpeeding) [reply pipelines/config.py:98]: Btw --ignore-words-list doesn't exist rn, that's in this pr https://github.com/codespell-project/codespell/pull/2040. You'd probably want to use --ignore-words and make a file for it rn

**Comment by FasterSpeeding** [comment]
was this never merged? I thought this was a new pr lol

File: hikari/api/rest.py [hunk 7]
```diff
@@ -5097,21 +5052,18 @@ async def edit_my_nick(
     ) -> None:
         """Edit the associated token's member nick.
 
-        .. deprecated:: 2.0.0.dev104
-            Use `RESTClient.edit_my_member`'s `nick` argument instead.
```
Comment (FasterSpeeding) [inline hikari/api/rest.py:5101]: Why was this removed?

File: hikari/internal/deprecation.py [hunk 8]
```diff
@@ -35,72 +35,63 @@
...
-        )
-        obj.__doc__ = doc
+        old_doc = inspect.getdoc(obj)
+
+        # If the docstring is inherited we can assume that the deprecation warning was already added there
+        if old_doc:
+            first_line_end = old_doc.index("\n")
+            obj.__doc__ = (
+                old_doc[:first_line_end]
+                + f"\n\n.. deprecated:: {version}\n    {additional_information}"
```
Comment (FasterSpeeding) [inline hikari/internal/deprecation.py:88]: Wouldn't this lead to inconsistencies as to when the deprecation warnings are actually in the docs since it'll have to be done inline for cases where specific arguments were deprecated?

Comment (davfsa) [reply hikari/internal/deprecation.py:88]: This is aimed for fully deprecated functions/methods/classes. When deprecating only an argument we can add the warning to the docstring and then raise the warning with `warn_deprecated`

Comment (FasterSpeeding) [reply hikari/internal/deprecation.py:88]: Which makes the docs inconsistent for anybody reading through the code right?

Comment (davfsa) [reply hikari/internal/deprecation.py:88]: Not really, since the decorator would be there. The deprecation decorator could also be extended to include arguments and detect when a specific argument is passed

File: hikari/internal/deprecation.py [hunk 9]
```diff
@@ -35,72 +35,63 @@
...
-    version: typing.Optional[str]
-        If specified, the version it will be removed in.
-    alternative: typing.Optional[str]
-        If specified, the alternative to use.
     stack_level: int
         The stack level for the warning. Defaults to `3`.
     """
-    if inspect.isclass(obj) or inspect.isfunction(obj):
+    if inspect.isclass(obj):
+        action = ("Instantiation of", "class")
```
Comment (FasterSpeeding) [inline hikari/internal/deprecation.py:54]: Also this isn't fit for classes as it'd break anything inheriting from said class so this seems a bit redundant

File: hikari/internal/deprecation.py [hunk 10]
```diff
@@ -35,72 +35,63 @@
...
-        doc = inspect.getdoc(obj) or ""
-        doc += (
-            "\n"
-            "!!! warning\n"
-            f"    This {type_str} is deprecated and will be removed in {version_str}.\n"
-            f"    {alternative_str}\n"
-        )
-        obj.__doc__ = doc
+        old_doc = inspect.getdoc(obj)
+
```
Comment (FasterSpeeding) [inline hikari/internal/deprecation.py:82]: can we get a check here to error if this ever decorates a class

**Review by FasterSpeeding — APPROVED** [review]

**Comment by davfsa** [comment]
Finally done with all points from #808, so it can be considered closed and solved for the most part. Will be actually closed once this is merged into master :)
